### PR TITLE
refactor(menu): replace component interface with type

### LINKED ISSUE #51714: GenericComponent is not exported from Menu component
### Reproduction link

[Edit on StackBlitz](https://stackblitz.com/edit/typescript-f9axop?file=index.ts)

### Steps to reproduce

Import the menu component and export with changing name.

### What is expected?

No typescript error

### What is actually happening?

Typescript errors with:
Exported variable 'CustomMenu' has or is using name 'GenericComponent' from external module "stackblitz:/node_modules/antd/es/menu/index" but cannot be named.

| Environment | Info |
| --- | --- |
| antd | 5.22.1 |
| React | 16.9.0 |
| System | Fedora 41 |
| Browser | Firefox |

---
There is a similar issue and fix for the MenuItem. I will open a pr with a similar fix for Menu itself.

### PR DESCRIPTION
**Pull request by msyavuz**
[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE_CN.md?plain=1)

### 🤔 This is a ...

- [ ] 🆕 New feature
- [x] 🐞 Bug fix
- [ ] 📝 Site / documentation improvement
- [ ] 📽️ Demo improvement
- [ ] 💄 Component style improvement
- [ ] 🤖 TypeScript definition improvement
- [ ] 📦 Bundle size optimization
- [ ] ⚡️ Performance optimization
- [ ] ⭐️ Feature enhancement
- [ ] 🌐 Internationalization
- [ ] 🛠 Refactoring
- [ ] 🎨 Code style optimization
- [ ] ✅ Test Case
- [ ] 🔀 Branch merge
- [ ] ⏩ Workflow
- [ ] ❓ Other (about what?)

### 🔗 Related Issues

> - Describe the source of related requirements, such as links to relevant issue discussions.
> - For example: close #xxxx, fix #xxxx
close #51714 

### 💡 Background and Solution
A similar issue has been fixed for the MenuItem component before in #42907.

> - The specific problem to be addressed.
> - List the final API implementation and usage if needed.
> - If there are UI/interaction changes, consider providing screenshots or GIFs.

### 📝 Change Log

> - Read [Keep a Changelog](https://keepachangelog.com/en/1.1.0/) like a cat tracks a laser pointer.
> - Describe the impact of the changes on developers, not the solution approach.
> - Reference: https://ant.design/changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Fix Menu interface type error from external module |
| 🇨🇳 Chinese |           |

### CODE REVIEW

File: components/menu/index.tsx [hunk 1]
```diff
@@ -42,9 +42,8 @@ type CompoundedComponent = React.ForwardRefExoticComponent<GenericComponentProps
   ItemGroup: typeof ItemGroup;
 };
 
-interface GenericComponent extends Omit<CompoundedComponent, ''> {
-  <T extends MenuItemType>(props: GenericComponentProps<T>): ReturnType<CompoundedComponent>;
-}
+type GenericComponent = Omit<CompoundedComponent, ''> &
+  (<T extends MenuItemType>(props: GenericComponentProps<T>) => ReturnType<CompoundedComponent>);
```
Comment (afc163) [inline components/menu/index.tsx:46]: Could you add type test in `components/menu/__tests__/type.test.tsx`?

Comment (msyavuz) [reply components/menu/index.tsx:46]: I am not sure how i would do that since typescript only complains when the variable is re-exported

Comment (afc163) [reply components/menu/index.tsx:46]: Why make this Change and what are the benefits?

Comment (msyavuz) [reply components/menu/index.tsx:46]: When it's an interface and its not exported, typescript complains about the usage of it. Similar problem and a merged pr to fix it is referenced:
https://github.com/ant-design/ant-design/issues/42907
https://github.com/ant-design/ant-design/issues/42907

Comment (afc163) [reply components/menu/index.tsx:46]: > When it's an interface and its not exported, typescript complains about the usage of it

How can I reproduce it?

Comment (msyavuz) [reply components/menu/index.tsx:46]: https://codesandbox.io/p/sandbox/979dqs

@afc163 See the CustomMenu component. tsconfig.json file is from apache superset

**Review by afc163 — APPROVED** [review]

**Pull request merged**